File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/ui/calendar/times/SunView.java
package com.byagowi.persiancalendar.ui.calendar.times;

import android.animation.ArgbEvaluator;
import android.animation.ValueAnimator;
import android.annotation.TargetApi;
import android.content.Context;
import android.content.res.Resources;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.DashPathEffect;
import android.graphics.LinearGradient;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.PathEffect;
import android.graphics.RectF;
import android.graphics.Shader;
import android.os.Build;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.View;
import android.view.animation.DecelerateInterpolator;

import com.byagowi.persiancalendar.R;
import com.byagowi.persiancalendar.praytimes.Clock;
import com.byagowi.persiancalendar.praytimes.PrayTimes;
import com.byagowi.persiancalendar.utils.TypefaceUtils;
import com.byagowi.persiancalendar.utils.Utils;

import java.util.Calendar;
import java.util.Locale;

import androidx.annotation.ColorInt;
import androidx.core.content.ContextCompat;

/**
 * @author MEHDI DIMYADI
 * MEHDIMYADI
 */

public class SunView extends View implements ValueAnimator.AnimatorUpdateListener {

    private final float FULL_DAY = new Clock(24, 0).toInt();
    private final float HALF_DAY = new Clock(12, 0).toInt();
    Paint mPaint, mSunPaint, mSunRaisePaint, mDayPaint;
    @ColorInt
    int horizonColor, timelineColor, taggingColor, nightColor, dayColor, daySecondColor, sunColor,
            sunBeforeMiddayColor, sunAfterMiddayColor, sunEveningColor, sunriseTextColor,
            middayTextColor, sunsetTextColor, colorTextNormal, colorTextSecond;
    int width, height;
    Path curvePath, nightPath;
    float current = 0;
    LinearGradient linearGradient = new LinearGradient(0, 0, 1, 0, 0, 0, Shader.TileMode.MIRROR);
    Paint moonPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    Paint moonPaintB = new Paint(Paint.ANTI_ALIAS_FLAG);
    Paint moonPaintO = new Paint(Paint.ANTI_ALIAS_FLAG);
    Paint moonPaintD = new Paint(Paint.ANTI_ALIAS_FLAG);
    RectF moonRect = new RectF();
    RectF moonOval = new RectF();
    String dayLengthString = "";
    String remainingString = "";
    String sunriseString = "";
    String middayString = "";
    String sunsetString = "";
    boolean isRTL = false;
    private double segmentByPixel;
    private ArgbEvaluator argbEvaluator = new ArgbEvaluator();
    private PrayTimes prayTimes;
    //    private Horizontal moonPosition;
    private double moonPhase = 1;
    private int fontSize;

    public SunView(Context context) {
        super(context);

        init(context, null);
    }

    public SunView(Context context, AttributeSet attrs) {
        super(context, attrs);

        init(context, attrs);
    }

    public SunView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        init(context, attrs);
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    public SunView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);

        init(context, attrs);
    }

    private void init(Context context, AttributeSet attrs) {
        if (attrs != null) {
            TypedArray typedArray = context.obtainStyledAttributes(attrs, R.styleable.SunView);
            TypedValue typedValue = new TypedValue();
            Resources.Theme theme = context.getTheme();

            try {
                theme.resolveAttribute(R.attr.SunViewHorizonColor, typedValue, true);
                int HorizonColor = ContextCompat.getColor(context, typedValue.resourceId);
                horizonColor = typedArray.getColor(R.styleable.SunView_SunViewHorizonColor, HorizonColor);
                theme.resolveAttribute(R.attr.SunViewTimelineColor, typedValue, true);
                int TimelineColor = ContextCompat.getColor(context, typedValue.resourceId);
                timelineColor = typedArray.getColor(R.styleable.SunView_SunViewHorizonColor, TimelineColor);
                theme.resolveAttribute(R.attr.SunViewTaglineColor, typedValue, true);
                int taglineColor = ContextCompat.getColor(context, typedValue.resourceId);
                taggingColor = typedArray.getColor(R.styleable.SunView_SunViewHorizonColor, taglineColor);
                nightColor = typedArray.getColor(R.styleable.SunView_SunViewNightColor, ContextCompat.getColor(context, R.color.sViewNightColor));
                dayColor = typedArray.getColor(R.styleable.SunView_SunViewDayColor, ContextCompat.getColor(context, R.color.sViewDayColor));
                daySecondColor = typedArray.getColor(R.styleable.SunView_SunViewDaySecondColor, ContextCompat.getColor(context, R.color.sViewDaySecondColor));
                sunColor = typedArray.getColor(R.styleable.SunView_SunViewSunColor, ContextCompat.getColor(context, R.color.sViewSunColor));
                sunBeforeMiddayColor = typedArray.getColor(R.styleable.SunView_SunViewBeforeMiddayColor, ContextCompat.getColor(context, R.color.sViewSunBeforeMiddayColor));
                sunAfterMiddayColor = typedArray.getColor(R.styleable.SunView_SunViewAfterMiddayColor, ContextCompat.getColor(context, R.color.sViewSunAfterMiddayColor));
                sunEveningColor = typedArray.getColor(R.styleable.SunView_SunViewEveningColor, ContextCompat.getColor(context, R.color.sViewSunEveningColor));
                theme.resolveAttribute(R.attr.SunViewSunriseTextColor, typedValue, true);
                int SunriseTextColor = ContextCompat.getColor(context, typedValue.resourceId);
                sunriseTextColor = typedArray.getColor(R.styleable.SunView_SunViewSunriseTextColor, SunriseTextColor);
                theme.resolveAttribute(R.attr.SunViewMiddayTextColor, typedValue, true);
                int MiddayTextColor = ContextCompat.getColor(context, typedValue.resourceId);
                middayTextColor = typedArray.getColor(R.styleable.SunView_SunViewMiddayTextColor, MiddayTextColor);
                theme.resolveAttribute(R.attr.SunViewSunsetTextColor, typedValue, true);
                int SunsetTextColor = ContextCompat.getColor(context, typedValue.resourceId);
                sunsetTextColor = typedArray.getColor(R.styleable.SunView_SunViewSunsetTextColor, SunsetTextColor);

                theme.resolveAttribute(R.attr.colorTextNormal, typedValue, true);
                colorTextNormal = ContextCompat.getColor(context, typedValue.resourceId);
                theme.resolveAttribute(R.attr.colorTextSecond, typedValue, true);
                colorTextSecond = ContextCompat.getColor(context, typedValue.resourceId);

                fontSize = dpToPx(14);
            } finally {
                typedArray.recycle();
            }
        }

        mPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mPaint.setTypeface(TypefaceUtils.getAppFont(context));

        mSunPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mSunPaint.setColor(sunColor);
        mSunPaint.setStyle(Paint.Style.FILL);

        mSunRaisePaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mSunRaisePaint.setColor(sunColor);
        mSunRaisePaint.setStyle(Paint.Style.STROKE);
        mSunRaisePaint.setStrokeWidth(7);
        PathEffect sunRaysEffects = new DashPathEffect(new float[]{3, 7}, 0);
        mSunRaisePaint.setPathEffect(sunRaysEffects);

        mDayPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mDayPaint.setStyle(Paint.Style.FILL_AND_STROKE);
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldW, int oldH) {
        super.onSizeChanged(w, h, oldW, oldH);

        width = w;
        height = h - 18;

        curvePath = new Path();
        curvePath.moveTo(0, height);

        if (width != 0) {
            segmentByPixel = (2 * Math.PI) / width;
        }

        for (int x = 0; x <= width; x++) {
            curvePath.lineTo(x, getY(x, segmentByPixel, (int) (height * 0.9f)));
        }

        nightPath = new Path(curvePath);
        nightPath.setLastPoint(width, height);
        nightPath.lineTo(width, 0);
        nightPath.lineTo(0, 0);
        nightPath.close();
    }

    // https://stackoverflow.com/a/34763668
    private int dpToPx(int dp) {
        return (int) (dp * Resources.getSystem().getDisplayMetrics().density);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        canvas.save();

        // draw fill of night
        mPaint.setStyle(Paint.Style.FILL);
        mPaint.setColor(nightColor);
        canvas.clipRect(0, height * 0.75f, width * current, height);
        canvas.drawPath(nightPath, mPaint);

        canvas.restore();

        canvas.save();

        // draw fill of day
        canvas.clipRect(0, 0, width, height);
        canvas.clipRect(0, 0, width * current, height * 0.75f);
        mDayPaint.setShader(linearGradient);
        canvas.drawPath(curvePath, mDayPaint);

        canvas.restore();

        canvas.save();

        // draw time curve
        canvas.clipRect(0, 0, width, height);
        mPaint.setStrokeWidth(3);
        mPaint.setStyle(Paint.Style.STROKE);
        mPaint.setColor(timelineColor);
        canvas.drawPath(curvePath, mPaint);

        canvas.restore();

        // draw horizon line
        mPaint.setColor(horizonColor);
        canvas.drawLine(0, height * 0.75f, width, height * 0.75f, mPaint);

        // draw sunset and sunrise tag line indicator
        mPaint.setColor(taggingColor);
        mPaint.setStrokeWidth(2);
        canvas.drawLine(width * 0.17f, height * 0.3f, width * 0.17f, height * 0.7f, mPaint);
        canvas.drawLine(width * 0.83f, height * 0.3f, width * 0.83f, height * 0.7f, mPaint);
        canvas.drawLine(getWidth() / 2f, height * 0.7f, getWidth() / 2f, height * 0.8f, mPaint);

        // draw text
        mPaint.setTextAlign(Paint.Align.CENTER);
        mPaint.setTextSize(fontSize);
        mPaint.setStrokeWidth(0);
        mPaint.setStyle(Paint.Style.FILL);
        mPaint.setColor(sunriseTextColor);
        canvas.drawText(sunriseString, width * 0.17f, height * .2f, mPaint);
        mPaint.setColor(middayTextColor);
        canvas.drawText(middayString, width / 2f, height * .94f, mPaint);
        mPaint.setColor(sunsetTextColor);
        canvas.drawText(sunsetString, width * 0.83f, height * .2f, mPaint);

        // draw remaining time
        mPaint.setTextAlign(Paint.Align.CENTER);
        mPaint.setStrokeWidth(0);
        mPaint.setStyle(Paint.Style.FILL);
        mPaint.setColor(colorTextSecond);
        canvas.drawText(dayLengthString, width * (isRTL ? 0.70f : 0.30f), height * .94f, mPaint);
        if (!TextUtils.isEmpty(remainingString)) {
            canvas.drawText(remainingString, width * (isRTL ? 0.30f : 0.70f), height * .94f, mPaint);
        }

        // draw sun
        if (current >= 0.17f && current <= 0.83f) {

            @ColorInt
            int color = (int) argbEvaluator.evaluate(current,
                    sunBeforeMiddayColor, sunAfterMiddayColor);

            mSunPaint.setColor(color);
            //mSunRaisePaint.setColor(color);
            //mPaint.setShadowLayer(1.0f, 1.0f, 2.0f, 0x33000000);
            canvas.drawCircle(width * current, getY((int) (width * current), segmentByPixel, (int) (height * 0.9f)), (height * 0.09f), mSunPaint);
            //mPaint.clearShadowLayer();
            //canvas.drawCircle(width * current, getY((int) (width * current), segmentByPixel, (int) (height * 0.9f)), (height * 0.09f) - 5, mSunRaisePaint);
        } else {
            drawMoon(canvas);
        }
    }

    public void drawMoon(Canvas canvas) {
        // This is brought from QiblaCompassView with some modifications
        float r = (height * 0.08f);
        float radius = 1;
        float px = width * current;
        float py = getY((int) (width * current), segmentByPixel, (int) (height * 0.9f));
        moonPaint.reset();
        moonPaint.setFlags(Paint.ANTI_ALIAS_FLAG);
        moonPaint.setColor(Color.WHITE);
        moonPaint.setStyle(Paint.Style.FILL_AND_STROKE);
        moonPaintB.reset();// moon Paint Black
        moonPaintB.setFlags(Paint.ANTI_ALIAS_FLAG);
        moonPaintB.setColor(Color.BLACK);
        moonPaintB.setStyle(Paint.Style.FILL_AND_STROKE);
        moonPaintO.reset();// moon Paint for Oval
        moonPaintO.setFlags(Paint.ANTI_ALIAS_FLAG);
        moonPaintO.setColor(Color.WHITE);
        moonPaintO.setStyle(Paint.Style.FILL_AND_STROKE);
        moonPaintD.reset();// moon Paint for Diameter
        // draw
        moonPaintD.setColor(Color.GRAY);
        moonPaintD.setStyle(Paint.Style.STROKE);
        moonPaintD.setFlags(Paint.ANTI_ALIAS_FLAG);
        canvas.rotate(180, px, py);
        int eOffset = 0;
        // elevation Offset 0 for 0 degree; r for 90 degree
        moonRect.set(px - r, py + eOffset - radius - r, px + r, py + eOffset - radius + r);
        canvas.drawArc(moonRect, 90, 180, false, moonPaint);
        canvas.drawArc(moonRect, 270, 180, false, moonPaintB);
        int arcWidth = (int) ((moonPhase - 0.5) * (4 * r));
        moonPaintO.setColor(arcWidth < 0 ? Color.BLACK : Color.WHITE);
        moonOval.set(px - Math.abs(arcWidth) / 2f, py + eOffset - radius - r,
                px + Math.abs(arcWidth) / 2f, py + eOffset - radius + r);
        canvas.drawArc(moonOval, 0, 360, false, moonPaintO);
        canvas.drawArc(moonRect, 0, 360, false, moonPaintD);
        canvas.drawLine(px, py - radius, px, py + radius, moonPaintD);
        moonPaintD.setPathEffect(null);
    }

    private float getY(int x, double segment, int height) {
        double cos = (Math.cos(-Math.PI + (x * segment)) + 1) / 2;
        return height - (height * (float) cos) + (height * 0.1f);
    }

    public void setSunriseSunsetMoonPhase(PrayTimes prayTimes, double moonPhase) {
        this.prayTimes = prayTimes;
        this.moonPhase = moonPhase;
        postInvalidate();
    }

    public void startAnimate(boolean immediate) {
        Context context = getContext();
        if (prayTimes == null || context == null)
            return;

        isRTL = Utils.isRTL(context);
        sunriseString = context.getString(R.string.sunriseSunView);
        middayString = context.getString(R.string.middaySunView);
        sunsetString = context.getString(R.string.sunsetSunView);

        float sunset = prayTimes.getSunsetClock().toInt();
        float sunrise = prayTimes.getSunriseClock().toInt();
        float midnight = prayTimes.getMidnightClock().toInt();

        if (midnight > HALF_DAY) midnight = midnight - FULL_DAY;
        float now = new Clock(Calendar.getInstance(Locale.getDefault())).toInt();

        float c = 0;
        if (now <= sunrise) {
            if (sunrise != 0) {
                c = ((now - midnight) / sunrise) * 0.17f;
            }
        } else if (now <= sunset) {
            if (sunset - sunrise != 0) {
                c = (((now - sunrise) / (sunset - sunrise)) * 0.66f) + 0.17f;
            }
        } else {
            if (FULL_DAY + midnight - sunset != 0) {
                c = (((now - sunset) / (FULL_DAY + midnight - sunset)) * 0.17f) + 0.17f + 0.66f;
            }
        }

        Clock dayLength = Clock.fromInt((int) (sunset - sunrise));
        Clock remaining = Clock.fromInt(now > sunset || now < sunrise ? 0 : (int) (sunset - now));
        dayLengthString = String.format(context.getString(R.string.length_of_day),
                Utils.formatNumber(dayLength.getHour()),
                Utils.formatNumber(dayLength.getMinute()));
        if (remaining.toInt() == 0) {
            remainingString = "";
        } else {
            remainingString = String.format(context.getString(R.string.remaining_daylight),
                    Utils.formatNumber(remaining.getHour()),
                    Utils.formatNumber(remaining.getMinute()));
        }

        argbEvaluator = new ArgbEvaluator();

        linearGradient = new LinearGradient(getWidth() * 0.17f, 0, getWidth() * 0.5f, 0,
                dayColor, daySecondColor, Shader.TileMode.MIRROR);

        if (immediate) {
            current = c;
            postInvalidate();
        } else {
            ValueAnimator animator = ValueAnimator.ofFloat(0, c);
            animator.setDuration(1500L);
            animator.setInterpolator(new DecelerateInterpolator());
            animator.addUpdateListener(this);
            animator.start();
        }
    }

    public void clear() {
        current = 0;
        postInvalidate();
    }

    @Override
    public void onAnimationUpdate(ValueAnimator valueAnimator) {
        current = (float) valueAnimator.getAnimatedValue();
        postInvalidate();
    }
}

File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/ui/calendar/times/SunView.kt
package com.byagowi.persiancalendar.ui.calendar.times

import android.animation.ArgbEvaluator
import android.animation.ValueAnimator
import android.annotation.SuppressLint
import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.util.TypedValue
import android.view.View
import android.view.animation.DecelerateInterpolator
import androidx.annotation.ColorInt
import androidx.core.content.ContextCompat
import com.byagowi.persiancalendar.LANG_EN_IR
import com.byagowi.persiancalendar.LANG_EN_US
import com.byagowi.persiancalendar.LANG_JA
import com.byagowi.persiancalendar.R
import com.byagowi.persiancalendar.utils.*
import io.github.persiancalendar.praytimes.Clock
import io.github.persiancalendar.praytimes.PrayTimes
import java.util.*
import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.cos

/**
 * @author MEHDI DIMYADI
 * MEHDIMYADI
 */

class SunView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null) :
        View(context, attrs), ValueAnimator.AnimatorUpdateListener {

    private val fullDay = Clock(24, 0).toInt().toFloat()
    private val halfDay = Clock(12, 0).toInt().toFloat()

    private val paint = Paint(Paint.ANTI_ALIAS_FLAG).also { it.typeface = getAppFont(context) }
    private val sunPaint = Paint(Paint.ANTI_ALIAS_FLAG).also { it.style = Paint.Style.FILL }
    private var sunRaisePaint = Paint(Paint.ANTI_ALIAS_FLAG).also {
        it.style = Paint.Style.STROKE
        it.strokeWidth = 7f
        it.pathEffect = DashPathEffect(floatArrayOf(3f, 7f), 0f) /* Sun rays effect */
    }
    private val dayPaint =
            Paint(Paint.ANTI_ALIAS_FLAG).also { it.style = Paint.Style.FILL_AND_STROKE }

    private var horizonColor: Int = 0
    private var timelineColor: Int = 0
    private var taggingColor: Int = 0
    private var nightColor = ContextCompat.getColor(context, R.color.sViewNightColor)
    private var dayColor = ContextCompat.getColor(context, R.color.sViewDayColor)
    private var daySecondColor = ContextCompat.getColor(context, R.color.sViewDaySecondColor)
    private var sunColor = ContextCompat.getColor(context, R.color.sViewSunColor)
    private var sunBeforeMiddayColor = ContextCompat.getColor(context, R.color.sViewSunBeforeMiddayColor)
    private var sunAfterMiddayColor = ContextCompat.getColor(context, R.color.sViewSunAfterMiddayColor)
    private var sunriseTextColor: Int = 0
    private var middayTextColor: Int = 0
    private var sunsetTextColor: Int = 0
    private var colorTextNormal: Int = 0
    private var colorTextSecond: Int = 0

    internal var width: Int = 0
    internal var height: Int = 0
    private lateinit var curvePath: Path
    private lateinit var nightPath: Path
    private var current = 0f
    private var moonPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private var moonPaintB = Paint(Paint.ANTI_ALIAS_FLAG)
    private var moonPaintO = Paint(Paint.ANTI_ALIAS_FLAG)
    private var moonPaintD = Paint(Paint.ANTI_ALIAS_FLAG)
    private var moonRect = RectF()
    private var moonOval = RectF()
    private var dayLengthString = ""
    private var remainingString = ""
    private var sunriseString = ""
    private var middayString = ""
    private var sunsetString = ""
    private var isRTL = false
    private var isShaderInitiationNeeded = true
    private var segmentByPixel = .0
    private var argbEvaluator = ArgbEvaluator()
    private var prayTimes: PrayTimes? = null

    //    private Horizontal moonPosition;
    private var moonPhase = 1.0
    private var fontSize: Int = 0

    init {
        val tempTypedValue = TypedValue()

        @ColorInt
        fun resolveColor(attr: Int) = tempTypedValue.let {
            context.theme.resolveAttribute(attr, it, true)
            ContextCompat.getColor(context, it.resourceId)
        }

        colorTextNormal = resolveColor(R.attr.colorTextNormal)
        colorTextSecond = resolveColor(R.attr.colorTextSecond)

        horizonColor = resolveColor(R.attr.SunViewHorizonColor)
        timelineColor = resolveColor(R.attr.SunViewTimelineColor)
        taggingColor = resolveColor(R.attr.SunViewTaglineColor)
        sunriseTextColor = resolveColor(R.attr.SunViewSunriseTextColor)
        middayTextColor = resolveColor(R.attr.SunViewMiddayTextColor)
        sunsetTextColor = resolveColor(R.attr.SunViewSunsetTextColor)

        // resolveColor(R.attr.SunViewNightColor)
        // resolveColor(R.attr.SunViewDayColor)
        // resolveColor(R.attr.SunViewDaySecondColor)
        // resolveColor(R.attr.SunViewSunColor)
        // resolveColor(R.attr.SunViewBeforeMiddayColor)
        // resolveColor(R.attr.SunViewAfterMiddayColor)
        // resolveColor(R.attr.SunViewEveningColor)

        sunPaint.color = sunColor
        sunRaisePaint.color = sunColor

        fontSize = when (language) {
            LANG_EN_IR, LANG_EN_US, LANG_JA -> {
                12.dp
            }
            else -> 14.dp
        }
    }

    override fun onSizeChanged(w: Int, h: Int, oldW: Int, oldH: Int) {
        super.onSizeChanged(w, h, oldW, oldH)

        isShaderInitiationNeeded = true

        width = w
        height = h - 18

        curvePath = Path()
        curvePath.moveTo(0f, height.toFloat())

        when {
            width != 0 -> {
                segmentByPixel = 2 * PI / width
            }
        }

        (0..width).forEach {
            curvePath.lineTo(it.toFloat(), getY(it, segmentByPixel, (height * .9f).toInt()))
        }

        nightPath = Path(curvePath).also {
            it.setLastPoint(width.toFloat(), height.toFloat())
            it.lineTo(width.toFloat(), 0f)
            it.lineTo(0f, 0f)
            it.close()
        }
    }

    @SuppressLint("DrawAllocation")
    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        val width = width
        val height = height
        val isRTL = isRTL

        when {
            isShaderInitiationNeeded -> {
                isShaderInitiationNeeded = false
                handler.postDelayed({
                    LinearGradient(
                            width * .17f,
                            0f,
                            width / 2f,
                            0f,
                            dayColor,
                            daySecondColor,
                            Shader.TileMode.MIRROR
                    ).also { dayPaint.shader = it }
                    postInvalidate()
                }, 80)
            }
        }

        canvas.also {
            it.save()
            when {
                isRTL -> it.scale(-1f, 1f, width / 2f, height / 2f)
            }
            it.save()
        }

        // draw fill of night
        paint.also {
            it.style = Paint.Style.FILL
            it.color = nightColor
        }
        canvas.also {
            it.clipRect(0f, height * .75f, width * current, height.toFloat())
            it.drawPath(nightPath, paint)
            it.restore()
            it.save()
        }

        // draw fill of day
        canvas.also {
            it.clipRect(0, 0, width, height)
            it.clipRect(0f, 0f, width * current, height * .75f)
            it.drawPath(curvePath, dayPaint)
            it.restore()
            it.save()
        }

        // draw time curve
        canvas.clipRect(0, 0, width, height)
        paint.also {
            it.strokeWidth = 3f
            it.style = Paint.Style.STROKE
            it.color = timelineColor
        }
        canvas.also {
            it.drawPath(curvePath, paint)
            it.restore()
        }

        // draw horizon line
        paint.color = horizonColor
        canvas.drawLine(0f, height * .75f, width.toFloat(), height * .75f, paint)

        // draw sunset and sunrise tag line indicator
        paint.also {
            it.color = taggingColor
            it.strokeWidth = 2f
        }
        canvas.also {
            it.drawLine(width * .17f, height * .3f, width * .17f, height * .7f, paint)
            it.drawLine(width * .83f, height * .3f, width * .83f, height * .7f, paint)
            it.drawLine(width / 2f, height * .7f, width / 2f, height * .8f, paint)
        }

        // draw sun
        when (current) {
            in .17f..0.83f -> {

                @ColorInt
                val color = argbEvaluator.evaluate(
                        current,
                        sunBeforeMiddayColor, sunAfterMiddayColor
                ) as Int

                sunPaint.color = color
                //mSunRaisePaint.setColor(color);
                //mPaint.setShadowLayer(1.0f, 1.0f, 2.0f, 0x33000000);
                canvas.drawCircle(
                        width * current,
                        getY((width * current).toInt(), segmentByPixel, (height * .9f).toInt()),
                        height * .09f,
                        sunPaint
                )
                //mPaint.clearShadowLayer();
                //canvas.drawCircle(width * current, getY((int) (width * current), segmentByPixel, (int) (height * .9f)), (height * .09f) - 5, mSunRaisePaint);
            }
            else -> {
                drawMoon(canvas)
            }
        }

        canvas.restore()

        // draw text
        paint.also {
            it.textAlign = Paint.Align.CENTER
            it.textSize = fontSize.toFloat()
            it.strokeWidth = 0f
            it.style = Paint.Style.FILL
            it.color = sunriseTextColor
        }
        canvas.drawText(
                sunriseString, width * when {
            isRTL -> .83f
            else -> .17f
        }, height * .2f, paint
        )
        paint.color = middayTextColor
        canvas.drawText(middayString, width / 2f, height * .94f, paint)
        paint.color = sunsetTextColor
        canvas.drawText(
                sunsetString, width * when {
            isRTL -> .17f
            else -> .83f
        }, height * .2f, paint
        )

        // draw remaining time
        paint.also {
            it.textAlign = Paint.Align.CENTER
            it.strokeWidth = 0f
            it.style = Paint.Style.FILL
            it.color = colorTextSecond
        }
        canvas.drawText(
                dayLengthString, width * when {
            isRTL -> .70f
            else -> .30f
        }, height * .94f, paint
        )
        when {
            remainingString.isNotEmpty() -> {
                canvas.drawText(
                        remainingString,
                        width * when {
                            isRTL -> .30f
                            else -> .70f
                        },
                        height * .94f,
                        paint
                )
            }
        }

    }

    private fun drawMoon(canvas: Canvas) {
        // This is brought from QiblaCompassView with some modifications
        val r = height * .08f
        val radius = 1f
        val px = width * current
        val py = getY((width * current).toInt(), segmentByPixel, (height * .9f).toInt())
        moonPaint.also {
            it.reset()
            it.flags = Paint.ANTI_ALIAS_FLAG
            it.color = Color.WHITE
            it.style = Paint.Style.FILL_AND_STROKE
        }
        moonPaintB.also {
            it.reset()// moon Paint Black
            it.flags = Paint.ANTI_ALIAS_FLAG
            it.color = Color.BLACK
            it.style = Paint.Style.FILL_AND_STROKE
        }
        moonPaintO.also {
            it.reset()// moon Paint for Oval
            it.flags = Paint.ANTI_ALIAS_FLAG
            it.color = Color.WHITE
            it.style = Paint.Style.FILL_AND_STROKE
        }
        moonPaintD.also {
            it.reset()// moon Paint for Diameter
            it.flags = Paint.ANTI_ALIAS_FLAG
            it.color = Color.GRAY
            it.style = Paint.Style.STROKE
        }
        canvas.rotate(180f, px, py)
        val eOffset = 0
        val arcWidth = ((moonPhase - .5) * (4 * r)).toInt()
        // elevation Offset 0 for 0 degree; r for 90 degree
        moonRect.set(px - r, py + eOffset - radius - r, px + r, py + eOffset - radius + r)
        canvas.also {
            it.drawArc(moonRect, 90f, 180f, false, moonPaint)
            it.drawArc(moonRect, 270f, 180f, false, moonPaintB)
        }
        moonPaintO.color = when {
            arcWidth < 0 -> Color.BLACK
            else -> Color.WHITE
        }
        moonOval.set(
                px - abs(arcWidth) / 2f, py + eOffset - radius - r,
                px + abs(arcWidth) / 2f, py + eOffset - radius + r
        )
        canvas.also {
            it.drawArc(moonOval, 0f, 360f, false, moonPaintO)
            it.drawArc(moonRect, 0f, 360f, false, moonPaintD)
            it.drawLine(px, py - radius, px, py + radius, moonPaintD)
        }
        moonPaintD.pathEffect = null
    }

    private fun getY(x: Int, segment: Double, height: Int): Float =
            height - height * ((cos(-PI + x * segment) + 1f) / 2f).toFloat() + height * .1f

    fun setSunriseSunsetMoonPhase(prayTimes: PrayTimes, moonPhase: Double) {
        this.prayTimes = prayTimes
        this.moonPhase = moonPhase
        postInvalidate()
    }

    fun startAnimate(immediate: Boolean = false) {
        val context = context ?: return
        val prayTimes = prayTimes ?: return

        isRTL = isRTL(context)
        sunriseString = context.getString(R.string.sunriseSunView)
        middayString = context.getString(R.string.middaySunView)
        sunsetString = context.getString(R.string.sunsetSunView)

        val sunset = prayTimes.sunsetClock.toInt().toFloat()
        val sunrise = prayTimes.sunriseClock.toInt().toFloat()
        var midnight = prayTimes.midnightClock.toInt().toFloat()

        when {
            midnight > halfDay -> midnight -= fullDay
        }
        val now = Clock(Calendar.getInstance(Locale.getDefault())).toInt().toFloat()

        var c = 0f
        when {
            now <= sunrise -> {
                when {
                    sunrise != 0f -> {
                        c = (now - midnight) / sunrise * .17f
                    }
                }
            }
            now <= sunset -> {
                when {
                    sunset - sunrise != 0f -> {
                        c = (now - sunrise) / (sunset - sunrise) * .66f + .17f
                    }
                }
            }
            else -> {
                when {
                    fullDay + midnight - sunset != 0f -> {
                        c = (now - sunset) / (fullDay + midnight - sunset) * .17f + .17f + .66f
                    }
                }
            }
        }

        val dayLength = Clock.fromInt((sunset - sunrise).toInt())
        val remaining =
                Clock.fromInt(
                        when {
                            now > sunset || now < sunrise -> 0
                            else -> (sunset - now).toInt()
                        }
                )
        dayLengthString = context.getString(R.string.length_of_day).format(
                formatNumber(dayLength.hour), formatNumber(dayLength.minute)
        )
        remainingString = when {
            remaining.toInt() == 0 -> ""
            else -> context.getString(
                    R.string.remaining_daylight
            ).format(formatNumber(remaining.hour), formatNumber(remaining.minute))
        }

        argbEvaluator = ArgbEvaluator()

        when {
            immediate -> {
                current = c
                postInvalidate()
            }
            else -> {
                ValueAnimator.ofFloat(0F, c).apply {
                    duration = 1500L
                    interpolator = DecelerateInterpolator()
                    addUpdateListener(this@SunView)
                    start()
                }
            }
        }
    }

    fun clear() {
        current = 0f
        postInvalidate()
    }

    override fun onAnimationUpdate(valueAnimator: ValueAnimator) {
        current = valueAnimator.animatedValue as Float
        postInvalidate()
    }
}
